### PR TITLE
Add integration tests for the CLI

### PR DESCRIPTION
**Pull request by ecton**
This change refactors how database tests are run. There are two
initialization methods that return a lock guard that controls the flow
of other database tests. Safe tests are written using a transaction to
isolate all of the database calls, and are therefore safe to run
simultaneously as other safe tests. Unsafe tests are written in a way
that they try to clean themselves up, but the test environment will
automatically clear the database before each unsafe test.

The last change is that tests are now run in their own database, and
during the first database test's initialization, all migrations are
unwound and then performed to give the database a clean slate.

These tests mean that the database migrations themselves are now run as
part of the test harness, which is another added bonus.

As an added bonus, while testing this feature, I realized that we both missed that the Remove permission select statement was coded to add the permission (fixed in the other PR). Clearly, it is beneficial to have end-to-end tests.

### CODE REVIEW

**Comment by ecton** [comment]
I think I'm happy with this PR. The tests do not test the actual output of the command, just that the effects were made correctly. This is partially out of laziness, but it's more out of a reluctance to spend time testing something I'm not sure I'm satisfied with. The table output is "ok" but it's not necessarily what I want it to be in the long run. As such, I felt like it'd be a waste of time to try to put tests around the table generation.

By this virtue, the view-only commands are completely untested.

Given how much is in flux in this project, I'm hoping you'll accept that lack of testing on that surface, and focus on whether you like my approach or not.

File: .github/workflows/deploy.yml [hunk 1]
```diff
@@ -26,8 +26,9 @@ jobs:
           ref: ${{ matrix.branch }}
       - name: Audit
         run: |
-          cargo audit -D warnings --ignore RUSTSEC-2020-0016
+          cargo audit -D warnings --ignore RUSTSEC-2020-0016 --ignore RUSTSEC-2020-0095
```
Comment (daxpedda) [inline .github/workflows/deploy.yml:29]: We should just really replace that crate, `cargo-audit` itself proposes some alternatives:
- [`dissimilar`](https://crates.io/crates/dissimilar)
- [`similar`](https://crates.io/crates/similar)
- [`treediff`](https://crates.io/crates/treediff)
- [`diffus`](https://crates.io/crates/diffus)

I know `similar` was made a direct replacement because of the lack of maintenance.

File: database/src/lib.rs [hunk 2]
```diff
@@ -71,20 +71,64 @@ impl<T> SqlxResultExt<T> for Result<T, sqlx::Error> {
     }
 }
 
-#[cfg(test)]
-mod test_util {
+pub mod test_util {
```
Comment (daxpedda) [inline database/src/lib.rs:74]: Why did you remove `#[cfg(test)]` here?

File: database/src/schema/account.rs [hunk 3]
```diff
@@ -110,6 +110,21 @@ impl Account {
         .map(|_| ())
     }
 
+    // Delete an account
```
Comment (daxpedda) [inline database/src/schema/account.rs:113]:
```suggestion
    // Delete an account.
```
Grammar police arrived.

File: server/tests/permission-groups-cli.rs [hunk 4]
```diff
@@ -0,0 +1,80 @@
+use assert_cmd::prelude::*;
+use cosmicverge_shared::permissions::{AccountPermission, Permission, Service};
+use database::schema::{PermissionGroup, Statement};
+use std::{env, process::Command, str::FromStr};
+
+#[tokio::test]
+async fn permission_groups() -> anyhow::Result<()> {
+    database::test_util::initialize_exclusive_test().await;
```
Comment (daxpedda) [inline server/tests/permission-groups-cli.rs:8]: Integration tests never run in parallel anyway. They are always in their own process too.

File: database/src/schema/account.rs [hunk 5]
```diff
@@ -110,6 +110,21 @@ impl Account {
...
+    //
+    // Do not use this outside of testing. When we support account deletion,
+    // there will be a process involved to ensure we're complying with
+    // international privacy laws.
+    pub async fn delete<'e, E: sqlx::Executor<'e, Database = sqlx::Postgres>>(
```
Comment (daxpedda) [inline database/src/schema/account.rs:118]: Shouldn't we still unit test this while we are at it?

File: database/src/schema/permission_group.rs [hunk 6]
```diff
@@ -104,6 +105,16 @@ impl PermissionGroup {
             .execute(executor)
             .await.map(|_| ())
     }
+
+    pub async fn delete<'e, E: sqlx::Executor<'e, Database = sqlx::Postgres>>(
```
Comment (daxpedda) [inline database/src/schema/permission_group.rs:109]: Unit test it, as above.

File: database/src/schema/twitch_profile.rs [hunk 7]
```diff
@@ -15,4 +15,13 @@ impl TwitchProfile {
             username,
         ).execute(executor).await.map(|_| ())
     }
+    pub async fn delete<'e, E: sqlx::Executor<'e, Database = sqlx::Postgres>>(
```
Comment (daxpedda) [inline database/src/schema/twitch_profile.rs:18]: Unit test it, as above.

**Review by daxpedda — CHANGES_REQUESTED** [review]

File: .github/workflows/deploy.yml (hunk 1 above)
Comment (ecton) [reply .github/workflows/deploy.yml:29]: We aren't directly using difference. assert_cmd is using predicates which is using this crate. We'd have to replace assert_cmd, but given the popularity of that crate and the fact that it's still in the rust book and apparently maintained, I would prefer just letting this in since it only runs during unit tests.

Comment (daxpedda) [reply .github/workflows/deploy.yml:29]: That sounds fine then, I can do a PR at some point, low prio.

File: database/src/lib.rs (hunk 2 above)
Comment (ecton) [reply database/src/lib.rs:74]: Because crates aren't recompiled with test mode on in the same workspace. These utilities need to be available to cosmicverge-server. With `#[cfg(test)]` and `pub mod test_util` the symbol still isn't available to other crates.

File: server/tests/permission-groups-cli.rs (hunk 4 above)
Comment (ecton) [reply server/tests/permission-groups-cli.rs:8]: That's good to know, so while the guard isn't important in this situation, the initialization code is still critical, and the method in which it initializes it (full wipe) is important.

File: database/src/schema/account.rs (hunk 5 above)
Comment (ecton) [reply database/src/schema/account.rs:118]: I mean, it's being included in the test surface, and it's not meant to be used in production, so I didn't see a need to unit test functionality that is only meant to be called within a test.

Unfortunately because of the same `#[cfg(test)]` limitation I described in the other comment, I can't make this method only available in tests. That's what I would prefer to do.

File: database/src/lib.rs (hunk 2 above)
Comment (daxpedda) [reply database/src/lib.rs:74]: Hm, hide it behind a feature called `test` and then add in `dev-dependencies` `database = { path = "../database", features = ["test"] }`.
I believe if we built some more extensive end-to-end tests in the future we are gonna need that anyway.

File: database/src/schema/permission_group.rs (hunk 6 above)
Comment (ecton) [reply database/src/schema/permission_group.rs:109]: This was the same argument as above, originally I was trying to gate it behind `#[cfg(test)]`, but unlike above, this function will be used eventually, so I will add tests for this (and the other one, and honestly, I'll just add it for Acocunt too at this point lol)

File: database/src/schema/account.rs (hunk 3 above)
Comment (daxpedda) [reply database/src/schema/account.rs:113]: Alright, don't mind me :smile:.
I was thinking that either all documentation should be full sentences or they shouldn't. As a german speaking person I feel uncomfortable around sentence fragments.

But you are the better judge of these kinda things, we aren't writing in german after all.

File: server/tests/permission-groups-cli.rs (hunk 4 above)
Comment (daxpedda) [reply server/tests/permission-groups-cli.rs:8]: Sounds good.

File: database/src/schema/account.rs (hunk 5 above)
Comment (daxpedda) [reply database/src/schema/account.rs:118]: > Do not use this outside of testing.

Yes, as you said in your own comment. Sry.

We can do the feature thing I described above to work around that issue.

File: database/src/schema/permission_group.rs (hunk 6 above)
Comment (daxpedda) [reply database/src/schema/permission_group.rs:109]: As above, but I think if this is only supposed to be called in tests, then let's not unit test it, thats kinda ridiculous :smile:.

File: database/src/lib.rs (hunk 2 above)
Comment (ecton) [reply database/src/lib.rs:74]: These functions need to be available to other crates, and `#[cfg(test)]` prevents the module from being exported.

File: database/src/schema/account.rs (hunk 3 above)
Comment (ecton) [reply database/src/schema/account.rs:114]: This is a sentence fragment, thus no period. A full sentence would look like "This function deletes an account" or "delete() removes an account from the database". I feel like those are too verbose, so I vote no period :)

File: server/tests/permission-groups-cli.rs (hunk 4 above)
Comment (ecton) [reply server/tests/permission-groups-cli.rs:8]: This is good to know, but this call still needs to be here. It wipes the database clean and ensures the database is initialized, both which are required by the idea of what an "exclusive test" needs.

File: database/src/schema/account.rs (hunk 5 above)
Comment (ecton) [reply database/src/schema/account.rs:119]: Going to wrap this one in the feature flag, but the others will be unit tested.

File: database/src/schema/account.rs (hunk 3 above)
Comment (ecton) [reply database/src/schema/account.rs:114]: I'm going to lean towards the more brief summaries for now. I feel like when you're reading the documentation, the subject is already implied as being the item you're documenting, so I don't like the idea of repeating or referring to it in the summary because it's redundant information. For that reason, I prefer a fragment as a "summary" of the item, usually followed by full sentences documenting it, if needed.

Let's agree to be a bit fluid early on with this. We haven't enabled the warning to document everything yet -- let's settle on a style when we do. In your next PR, try to document an item the way you normally would and we'll compare.

Comment (daxpedda) [reply database/src/schema/account.rs:114]: Sounds good.

File: database/src/schema/account.rs [hunk 8]
```diff
@@ -111,6 +111,22 @@ impl Account {
         .map(|_| ())
     }
 
+    // Delete an account
+    //
+    // Do expose this to be used outside of testing. When we support account deletion,
```
Comment (daxpedda) [inline database/src/schema/account.rs:116]:
```suggestion
    // Don't expose this to be used outside of testing. When we support account deletion,
```
Right? What happened here?

File: server/src/main.rs [hunk 9]
```diff
@@ -70,12 +79,15 @@ async fn main() -> anyhow::Result<()> {
...
-        Cli::Serve => http::run_webserver().await,
-        Cli::GenerateAssets { static_folder } => generate_assets(static_folder).await,
-        Cli::Account(command) => command.execute().await,
-        Cli::PermissionGroup(command) => command.execute().await,
+    let Cli {
+        database_url,
+        command,
+    } = Cli::from_args();
+    match command {
+        Command::Serve => http::run_webserver(database_url).await,
```
Comment (daxpedda) [inline server/src/main.rs:87]: I'm not sure if we are there yet, but do we want to rename this to API server?
We could also do this in a different PR.

**Review by daxpedda — CHANGES_REQUESTED** [review]

File: server/src/main.rs (hunk 9 above)
Comment (ecton) [reply server/src/main.rs:87]: Probably different PR, because there will still need to be an HTTP interface too, so I'm not sure exactly how everything will be spun up.

File: migrations/src/main.rs [hunk 10]
```diff
@@ -1,9 +1,10 @@
 mod connection;
+#[allow(dead_code)] // this mod exposes functions in lib.rs, but appears unused in the exe
```
[the remark this answers is not in the document]
Comment (ecton) [reply migrations/src/main.rs:2]: No, `undo_all` is never used in this executable. It's only used in testing right now.

I'm not really of a mindset that you ever roll back migrations in production, you're more likely to fix the issue related to what's affecting production, which usually means a new migration. The migrations executable thus only worries about bringing you up to date, and there's two functions available to roll back migrations during development: [debug](https://docs.rs/sqlx-simple-migrator/0.0.5/sqlx_simple_migrator/struct.Migration.html#method.debug) and [nuclear_debug](https://docs.rs/sqlx-simple-migrator/0.0.5/sqlx_simple_migrator/struct.Migration.html#method.nuclear_debug), both which are disabled when building for release.

[the remark this answers is not in the document]
Comment (daxpedda) [reply migrations/src/main.rs:2]: If it's only used in testing right now, why can't it hide behind the `test_util` feature?

**Review by daxpedda — APPROVED** [review]

**Review by daxpedda — APPROVED** [review]
